src/bindings/fluid_ladspa.h

```diff
@@ -21,220 +21,129 @@
 
-/* Purpose:
- * Applies the master gain (from command line option --gain or gain command).
- * Processes one block of sound data (generated from the synthesizer) through
- * the LADSPA Fx unit.
- * Acknowledges a bypass request.
- */
-void fluid_LADSPA_run(fluid_LADSPA_FxUnit_t* Fx_unit, fluid_real_t* left_buf[], fluid_real_t* right_buf[], fluid_real_t* fx_left_buf[], fluid_real_t* fx_right_buf[]);
+typedef enum _fluid_ladspa_node_type_t {
+    FLUID_LADSPA_NODE_AUDIO,
+    FLUID_LADSPA_NODE_CONTROL,
 
-/* Purpose:
- * Returns the node belonging to Name or NULL, if not found
- */
-fluid_LADSPA_Node_t* fluid_LADSPA_RetrieveNode(fluid_LADSPA_FxUnit_t* FxUnit, char * Name);
+} fluid_ladspa_node_type_t;
 
-/* Purpose:
- * Creates a new node with the given characteristics.
- */
-fluid_LADSPA_Node_t* fluid_LADSPA_CreateNode(fluid_LADSPA_FxUnit_t* FxUnit, char * Name, int flags);
-
-/* Purpose:
- * - Resets LADSPA Fx unit to bypass.
- * - Removes all plugins from the reverb unit.
- * - Releases all libraries.
- * Note: It would be more efficient to keep the libraries. But then the user would have to restart fluidsynth each time
- * a plugin is recompiled.
- */
-void fluid_LADSPA_clear(fluid_LADSPA_FxUnit_t* FxUnit);
+typedef struct _fluid_ladspa_lib_t
+{
+    char *filename;
+    void *dlib;
+    LADSPA_Descriptor_Function descriptor;
 
-/* Purpose:
- * Frees all memory and shuts down the Fx block.
- * The synthesis thread must be stopped, when calling.
- */
-void fluid_LADSPA_shutdown(fluid_LADSPA_FxUnit_t* FxUnit);
+} fluid_ladspa_lib_t;
 
+typedef struct _fluid_ladspa_port_state_t
+{
+    int num_inputs;
+    int num_outputs;
 
+} fluid_ladspa_port_state_t;
 
-/*
- * fluid_handle_LADSPA_XXX
- * Those functions are called from fluid_cmd, when a command is entered on the command line.
- */
+typedef struct _fluid_ladspa_plugin_t
+{
+    /* plugin instance id unique to the synth */
+    int id;
 
-/* Purpose:
- * - Resets LADSPA Fx unit to bypass.
- * - Removes all plugins from the reverb unit.
- * - Releases all libraries.
- * Note: It would be more efficient to keep the libraries. But then the user would have to restart fluidsynth each time
- * a plugin is recompiled.
- */
-int fluid_LADSPA_handle_clear(fluid_synth_t* synth, int ac, char** av, fluid_ostream_t out);
-
-/* Purpose:
- * Loads the plugins added with 'ladspa_add' and then start the Fx unit.
- * Internal processes:
- * - load the LADSPA plugin libraries
- * - instantiate the plugins
- * - connect the plugins
- * - set the bypass switch to 'not bypassed'
- */
-int fluid_LADSPA_handle_start(fluid_synth_t* synth, int ac, char** av, fluid_ostream_t out);
+    const LADSPA_Descriptor *desc;
+    LADSPA_Handle *handle;
 
-/* Purpose:
- * Adds one plugin into the list of the LADSPA Fx unit.
- * This is only allowed, while the Fx block is in 'bypass' state (after clear).
- */
-int fluid_LADSPA_handle_add(fluid_synth_t* synth, int ac, char** av, fluid_ostream_t out);
+    volatile int active;
```

Comment:
why volatile?

Reply:
To ensure this variable is guaranteed to be updated before the API mutex is released. Otherwise there could be a race condition, leading to a plugin being activated or deactivated more than once (which is forbidden in the LADSPA specs).

Reply:
You're always accessing it via a structure. I think it should be updated properly. However if you want guarantees you should use atomic setters to be sure.

Reply:
Sorry, wrong idea and wrong explanation. What I really want the volatile to protect the code against is weird compiler optimizations. The optimizer might try to store it in a local register and never re-read it again if it's not declared volatile. That might happen on the while loop in fluid_ladspa_deactivate. And as far as I know, accessing a struct member as opposed to a "normal" variable makes no difference to the compiler... but I might be mistaken?

On the plugin->active I think you are right, it is unnecessary and I will remove it.

Reply:
Ok, then just add a small comment to `state`, pointing out that register optimization thingy.

Reply:
I still think `state` is not optimized to a register, because it's properly accessed via struct and the compiler cannot be sure that this memory isnt aliased, because it doesnt live in local scope.

But since it's performance irrelevant it probably doesnt hurt to be volatile.